### ISSUE TITLE
Investigate regression in `tests/record_benchmark.py`

### ISSUE THREAD

**Issue by biobootloader** [AbanteAI/mentat#154]
@waydegg noticed some of these aren't doing well. I ran a few times and saw that as well. Let's run several times and determine our current performance for each one and determine the most common causes of failure.

<img width="942" alt="image" src="https://github.com/AbanteAI/mentat/assets/128252497/18742766-53ca-4a15-b940-8a73a8f528d0">

**Comment by biobootloader** [comment]
Related: https://github.com/AbanteAI/mentat/issues/155

**Comment by PCSwingle** [comment]
I turned off codemap and put auto tokens to 0 and it immediately started working most of the time

**Comment by biobootloader** [comment]
> I turned off codemap and put auto tokens to 0 and it immediately started working. It failed both times I tried at first, and succeeded all 4 times I tried after turning off codemap and autotokens

it may not even realize which file it's supposed to be working in...

**Comment by biobootloader** [comment]
Two possible solutions to try:

First, from @granawkins : 

>How about a tag after the fname saying who added it, and an explanation in the prompt. e.g.:
>"Your input will consist of a user request, some user-included code files (filenames appended ' USER'), some >automatically selected files (appended ' AUTO'), and sometimes the git diff of current code files.
>...
>mentat/conversation.py AUTO
>\<cmap>
>mentat/code_context.py USER
>\<code>

Second:

Add something like a list of user-added files as part of the beginning prompt, like "The user has run the system, specifically requesting to include files A, B, C in your context. User requests may implicitly refer to these files or their content. Other files and information are added automatically by the system, and may be useful to you."

**Comment by PCSwingle** [comment]
I honestly think the best solution is to not have any auto-tokens or codemap by default; it increases the token size (and therefore cost), and it's going to distract it from what the user actually wanted it to focus on. Instead, I think we should focus on having the model determine what files it needs (via the codemap and possibly embeddings) in a separate query with separate conversation history (so nothing bleeds over to the main conversation) and then tell the user that the model added these files before we show it's actual result. The workflow would be something like this:

On mentat start: Get user provided files (multifile_calculator/calculator.py) -> user gives query (add exp and log to calculator) -> we run gpt with codemap (and/or embeddings) to determine files needed for this query (I need multifile_calculator/operations.py) -> Tell user 'Model added operations.py to context' -> Run and stream main model response (has no idea that operations.py was added by previous model, has no idea what a codemap is or embeddings are).

Let me know what you think @biobootloader and @granawkins

**Comment by biobootloader** [comment]
> The biggest 2 downsides here are obviously the extra time to process a separate model request and the cost of that model request.

Yeah, @granawkins has been planning to try using `gpt-3.5-turbo-16k` for this! it's 10x cheaper than `gpt-4` so could be viable

**Comment by biobootloader** [comment]
Thanks @PCSwingle for looking into this. Things to do:

- [ ] document exactly how it fails (can be a comment here). There were 3 tests that seemed to always fail, maybe run each twice and record what went wrong the 6 times. I.e. is it editing the wrong files? going completely off track? other issues?
- [ ] try to fix with the two methods posted in this above comment: https://github.com/AbanteAI/mentat/issues/154#issuecomment-1765044557. The second method might be faster to try.
- [ ] If both those methods don't work, disable code maps / added context, allowing users to turn on with a warning to specify the files they want changes made to clearly.

**Comment by granawkins** [comment]
> On mentat start: Get user provided files (multifile_calculator/calculator.py) -> user gives query (add exp and log to calculator) -> we run gpt with codemap (and/or embeddings) to determine files needed for this query (I need multifile_calculator/operations.py) -> Tell user 'Model added operations.py to context' -> Run and stream main model response (has no idea that operations.py was added by previous model and not user, has no idea what a codemap is or embeddings are).

Yes, I'm right with you [here](https://github.com/AbanteAI/mentat/blob/89f5875dd00ee75358aa326690aa07e7f1410677/tests/benchmarks/context_benchmark.py#L19C32-L19C32). 

> Thanks @PCSwingle for looking into this. Things to do:
> 
> * [ ]  document exactly how it fails (can be a comment here). There were 3 tests that seemed to always fail, maybe run each twice and record what went wrong the 6 times. I.e. is it editing the wrong files? going completely off track? other issues?
> * [ ]  try to fix with the two methods posted in this above comment: [Investigate regression in `tests/record_benchmark.py` #154 (comment)](https://github.com/AbanteAI/mentat/issues/154#issuecomment-1765044557). The second method might be faster to try.
> * [ ]  If both those methods don't work, disable code maps / added context, allowing users to turn on with a warning to specify the files they want changes made to clearly.

@PCSwingle If you post the test prompt/expected results, I can add it to `context_benchmark.py`. I'd also be interested to run it myself and dig into why it's failing since I'm spending so much time with context.

**Comment by biobootloader** [comment]
> I'd also be interested to run it myself and dig into why it's failing since I'm spending so much time with context.

Go for it @granawkins! It might fail for different reasons each time, so go ahead and run and post about it. We'll either get extra information or it'll save Paul some time. It'd be good to have more eyes on it.

**Comment by PCSwingle** [comment]
> * [ ]  document exactly how it fails (can be a comment here). There were 3 tests that seemed to always fail, maybe run each twice and record what went wrong the 6 times. I.e. is it editing the wrong files? going completely off track? other issues?

I looked into both test_calculator_add_exp_then_log and test_calculator_continue_change; for the first one, the vast majority of the time GPT just got confused because there was a multifile_calculator/calculator (and operations) and there was a scripts/calculator, so it kept trying to edit the multifile_calculator. For the second one, it just doesn't know what change there is to continue; and I agree with it. This is the system message it sees:
```
Diff References:
 "-" = HEAD (last commit)
 "+" = Active Changes

Code Files:

.gitignore

__init__.py

multifile_calculator/__init__.py

multifile_calculator/calculator.py
function
	calculate (operation, num1, num2)

multifile_calculator/operations.py
function
	add_numbers (a, b)
	divide_numbers (a, b)
	multiply_numbers (a, b)
	subtract_numbers (a, b)

scripts/calculator.py
1:import fire
2:
3:
4:def add_numbers(a, b):
5:    return a + b
6:
7:
8:def multiply_numbers(a, b):
9:    return a * b
10:
11:
12:def subtract_numbers(a, b):
13:    return a - b
14:
15:
16:def divide_numbers(a, b):
17:    return a / b
18:
19:
20:def exp_numbers(a, b):
21:    return a**b
22:
23:
24:def calculate(operation, num1, num2):
25:    result = None
26:
27:    if operation == "add":
28:        result = add_numbers(num1, num2)
29:    elif operation == "subtract":
30:        result = subtract_numbers(num1, num2)
31:    elif operation == "multiply":
32:        result = multiply_numbers(num1, num2)
33:    elif operation == "divide":
34:        result = divide_numbers(num1, num2)
35:    else:
36:        print("Invalid operation")
37:
38:    return result
39:
40:
41:if __name__ == "__main__":
42:    fire.Fire(calculate)
43:

scripts/calculator.py
function
	add_numbers (a, b)
	calculate (operation, num1, num2)
	divide_numbers (a, b)
	exp_numbers (a, b)
	multiply_numbers (a, b)
	subtract_numbers (a, b)

+def exp_numbers(a, b):
+    return a**b
+
+
scripts/echo.py
function
	echo (value: Any)
	echo_hardcoded ()

scripts/graph_class.py
Graph
	member
		__init__ (self, values, connections)
		dfs (self, goal)
class
	Graph
function
	main ()
```
It's not exactly clear that those 4 lines are the ones that were added, not to mention it doesn't make much sense that the diff is right next to a bunch of function definitions.

**Comment by biobootloader** [comment]
Changing the defaults may help, I'll have to run and check: https://github.com/AbanteAI/mentat/pull/175

**Comment by biobootloader** [comment]
@granawkins, @PCSwingle and I took a look at the code message Paul posted in the above message and there are a couple issues we found:

1) there's some duplication, i.e. `scripts/calculator.py` is included both as full code and as a codemap. Paul discovered this was due to a path issue, so the same files were being added multiple times under different paths (relative and absolute). It's fixed here: https://github.com/AbanteAI/mentat/pull/179

2) The git diff lines aren't included with the full code lines? Maybe I was misremembering but I thought we had them inline after your original experiments on how the model best understood diff lines here: https://github.com/AbanteAI/mentat/pull/86. Maybe it's a bug that they got separated, or was it on purpose as part of adding features?
